sips/sip-012/sip-012-cost-limits-network-upgrade.md

```diff
@@ -0,0 +1,674 @@
+improvements in the implementation of the MARF (see [issue #2869](https://github.com/blockstack/stacks-blockchain/issues/2869))
+we propose doubling the current limits on blocks:
+
+```rust
+pub const BLOCK_LIMIT_MAINNET: ExecutionCost = ExecutionCost {
+    write_length: 15_000_000, // roughly 15 mb
+    write_count: 15_500,
+    read_length: 100_000_000,
+    read_count: 15_500,
+    runtime: 5_000_000_000,
+};
+```
+
+### Changes to Static vs. Dynamic Tabulation of Costs
+
+The cost assessment in Clarity for most data-handling functions (e.g.,
+`map-get?`) use the static cost of the fetch rather than the dynamic
+cost. For more information, see [issue #2864](https://github.com/blockstack/stacks-blockchain/issues/2864)
+in the `stacks-blockchain` repository.
+
+There were two motivating reasons for doing this:
+
+* It makes static analysis of costs easier, because the cost assessed
+  at runtime would always use the declared size of the map entry.
+* It allows the cost to be assessed before running the operation.
+
+However, it's pretty easy to argue that these reasons aren't all that
+essential. For (1), static analysis is always going to overestimate
+anyways, so why not allow the real cost to match the actual runtime
+overhead more closely. For (2), allowing a single "speculative"
+evaluation before aborting a block due to cost overflow is not
+particularly burdensome to the network: the maximum size of an
+overread is a single Clarity value, 2MB.
+
+The benefit of using dynamic costs, however, could be great. Many
+contracts use patterns where potentially long lists are stored in data
+maps, but in practice the stored lists are relatively short.
+
+Because of this, we propose to use a dynamic cost for these assessments.
+
+### Bitcoin Support Indication
+
+For wallets that wish to submit a supporting vote for the activation of this SIP via a BTC transaction, they can submit a transaction in the following form:
+
+The first input to the Bitcoin operation _must_ consume a UTXO that is
+a standard P2MS over P2SH output _or_ a standard P2PKH output. This first
+input will be used to verify that the sender of the Bitcoin operation corresponds
+to a specific Stacks address.
+
+The Bitcoin transaction will include an `OP_RETURN` output for the first Bitcoin output:
+
+```
+            0      2  3       9
+            |------|--|-------|
+              XM    A  SIP-12
+```
+
+That is, the `OP_RETURN` output will be a 9-byte field with content equal to `XMASIP-12` (ascii encoded).
+
+# Activation
+
+The SIP will be considered Active once:
+
+* A new release of stacks-blockchain is available with the updated cost-limits and a mechanism to use the new cost-limits beyond a pre-determined Bitcoin block height
+* This new release is deployed by independent miners, as determined by the continued operation of the Stacks blockchain beyond the Bitcoin block height selected for the network-upgrade. 
+
```

Comment:
The Activation section does not describe how long the voting process is allowed to take.  Please put a reasonable deadline on it (i.e. no more than 3 months of Bitcoin blocks).  Like, if we don't get enough support by then, then this SIP should fail to activate.

The Activation section also does not consider miners as a stakeholder, even though they would be affected by the upgrade.  Please describe a protocol by which they can signal support (or rejection), and update the Activation section to require that both the Stackers and miners have signed off on the upgrade.